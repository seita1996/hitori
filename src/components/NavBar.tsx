import { motion } from "framer-motion";
import { AppView } from "../types";

interface NavBarProps {
  currentView: AppView;
  setView: (view: AppView) => void;
}

export function NavBar({ currentView, setView }: NavBarProps) {
  return (
    <header className="border-b border-gray-200 dark:border-gray-700 bg-white dark:bg-gray-800 shadow-sm">
      <nav className="flex justify-between items-center p-4 max-w-6xl mx-auto">
        <h1 className="text-2xl font-bold text-indigo-600 dark:text-indigo-400">Hitori</h1>
        <div className="flex space-x-3">
          <NavButton
            active={currentView === "post"}
            onClick={() => setView("post")}
            label="投稿"
          />
          <NavButton
            active={currentView === "timeline"}
            onClick={() => setView("timeline")}
            label="履歴"
          />
          <NavButton
            active={currentView === "settings"}
            onClick={() => setView("settings")}
            label="設定"
          />
        </div>
      </nav>
    </header>
  );
}

interface NavButtonProps {
  active: boolean;
  onClick: () => void;
  label: string;
}

function NavButton({ active, onClick, label }: NavButtonProps) {
  return (
    <motion.button
      whileHover={{ scale: 1.05 }}
      whileTap={{ scale: 0.95 }}
      onClick={onClick}
      className={`px-5 py-2 rounded-xl font-medium transition-all duration-200 ${
        active
          ? "bg-indigo-500 text-white shadow-sm"
          : "bg-gray-100 dark:bg-gray-700 text-gray-600 dark:text-gray-300 hover:bg-gray-200 dark:hover:bg-gray-600"
      }`}
    >
      {label}
    </motion.button>
  );
}
