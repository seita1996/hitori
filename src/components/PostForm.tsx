import { useState, useRef, useEffect } from "react";
import { motion } from "framer-motion";

interface PostFormProps {
  onSubmit: (content: string) => void;
  maxLength?: number;
}

export function PostForm({ onSubmit, maxLength = 280 }: PostFormProps) {
  const [content, setContent] = useState("");
  const textAreaRef = useRef<HTMLTextAreaElement>(null);

  // テキストエリアの高さを自動調整
  useEffect(() => {
    if (textAreaRef.current) {
      textAreaRef.current.style.height = "auto";
      textAreaRef.current.style.height = `${textAreaRef.current.scrollHeight}px`;
    }
  }, [content]);

  const handleSubmit = () => {
    if (content.trim() && content.length <= maxLength) {
      onSubmit(content);
      setContent("");
      if (textAreaRef.current) {
        textAreaRef.current.style.height = "auto";
      }
    }
  };

  const remainingChars = maxLength - content.length;
  const isOverLimit = remainingChars < 0;

  return (
    <div className="bg-surface-light dark:bg-surface-dark rounded-2xl shadow-soft-lg p-6">
      <textarea
        ref={textAreaRef}
        value={content}
        onChange={(e) => setContent(e.target.value)}
        placeholder="今、何を思う？"
        className="w-full p-4 bg-white dark:bg-gray-800 rounded-xl border border-gray-200 dark:border-gray-700 focus:outline-none focus:ring-2 focus:ring-primary-light dark:focus:ring-primary-dark focus:border-transparent resize-none text-gray-700 dark:text-gray-200"
        rows={3}
      />

      <div className="flex justify-between items-center mt-4">
        <motion.span
          animate={{
            color: isOverLimit 
              ? "#ef4444" 
              : remainingChars < 20 
                ? "#fbbf24" 
                : "#9ca3af",
          }}
          className="text-sm font-medium"
        >
          {remainingChars}
        </motion.span>

        <motion.button
          whileTap={{ scale: 0.95 }}
          whileHover={{ scale: 1.05 }}
          onClick={handleSubmit}
          disabled={isOverLimit || !content.trim()}
          className={`px-6 py-2 rounded-xl font-medium transition-all duration-200 ${
            isOverLimit || !content.trim()
              ? "bg-gray-200 dark:bg-gray-700 text-gray-500"
              : "bg-primary-DEFAULT hover:bg-primary-dark text-white shadow-sm"
          }`}
        >
          投稿
        </motion.button>
      </div>
    </div>
  );
}
