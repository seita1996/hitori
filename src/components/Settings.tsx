import { useState } from "react";
import { motion } from "framer-motion";
import { CloudProvider, SyncStatus } from "../types";

interface SettingsProps {
  cloudProvider: CloudProvider;
  onProviderChange: (provider: CloudProvider) => void;
  onSync: () => Promise<void>;
  syncStatus: SyncStatus;
}

export function Settings({
  cloudProvider,
  onProviderChange,
  onSync,
  syncStatus
}: SettingsProps) {
  const [isSyncing, setIsSyncing] = useState(false);

  const handleSync = async () => {
    setIsSyncing(true);
    await onSync();
    setIsSyncing(false);
  };

  return (
    <div className="bg-white dark:bg-gray-800 rounded-2xl shadow-md p-6">
      <h2 className="text-xl font-semibold mb-6 text-gray-800 dark:text-gray-200">設定</h2>

      <div className="space-y-8">
        <div className="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700">
          <h3 className="text-lg font-medium mb-4 text-gray-700 dark:text-gray-300">クラウド同期</h3>
          <div className="space-y-3">
            <label className="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors">
              <input
                type="radio"
                checked={cloudProvider === "none"}
                onChange={() => onProviderChange("none")}
                className="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              />
              <span className="text-gray-700 dark:text-gray-300">同期しない</span>
            </label>
            <label className="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors">
              <input
                type="radio"
                checked={cloudProvider === "google"}
                onChange={() => onProviderChange("google")}
                className="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              />
              <span className="text-gray-700 dark:text-gray-300 flex items-center">
                Google Drive
              </span>
            </label>
            <label className="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors">
              <input
                type="radio"
                checked={cloudProvider === "apple"}
                onChange={() => onProviderChange("apple")}
                className="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              />
              <span className="text-gray-700 dark:text-gray-300 flex items-center">
                iCloud
              </span>
            </label>
          </div>
        </div>

        {cloudProvider !== "none" && (
          <motion.div
            initial={{ opacity: 0, height: 0 }}
            animate={{ opacity: 1, height: "auto" }}
            exit={{ opacity: 0, height: 0 }}
            className="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700"
          >
            <motion.button
              whileHover={{ scale: 1.03 }}
              whileTap={{ scale: 0.97 }}
              onClick={handleSync}
              disabled={isSyncing || syncStatus === "syncing"}
              className={`px-5 py-2.5 rounded-xl font-medium transition-all duration-200 ${
                isSyncing || syncStatus === "syncing"
                  ? "bg-gray-300 text-gray-700"
                  : "bg-indigo-500 text-white hover:bg-indigo-600 shadow-sm"
              }`}
            >
              <span className="flex items-center">
                <svg className={`w-5 h-5 mr-2 ${isSyncing ? "animate-spin" : ""}`} fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15" />
                </svg>
                {isSyncing ? "同期中..." : "今すぐ同期"}
              </span>
            </motion.button>

            <div className="mt-3 text-sm">
              {syncStatus === "synced" && (
                <span className="text-green-500 flex items-center">
                  <svg className="w-4 h-4 mr-1" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M5 13l4 4L19 7" />
                  </svg>
                  最後の同期: 成功
                </span>
              )}
              {syncStatus === "error" && (
                <span className="text-red-500 flex items-center">
                  <svg className="w-4 h-4 mr-1" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z" />
                  </svg>
                  同期エラー
                </span>
              )}
              {syncStatus === "offline" && (
                <span className="text-yellow-500 flex items-center">
                  <svg className="w-4 h-4 mr-1" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M18.364 5.636a9 9 0 010 12.728m0 0l-2.829-2.829m2.829 2.829L21 21M15.536 8.464a5 5 0 010 7.072m0 0l-2.829-2.829m-4.243 2.829a4.978 4.978 0 01-1.414-2.83m-1.414 5.658a9 9 0 01-2.167-9.238m7.824 2.167a1 1 0 111.414 1.414m-1.414-1.414L3 3m8.293 8.293l1.414 1.414" />
                  </svg>
                  オフライン
                </span>
              )}
            </div>
          </motion.div>
        )}

        <div className="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700">
          <h3 className="text-lg font-medium mb-4 text-gray-700 dark:text-gray-300">テーマ</h3>
          <button
            className="px-5 py-2.5 bg-gray-100 hover:bg-gray-200 dark:bg-gray-700 dark:hover:bg-gray-600 text-gray-700 dark:text-gray-300 rounded-xl transition-colors duration-200 flex items-center"
            onClick={() => {
              document.documentElement.classList.toggle("dark");
              localStorage.theme = document.documentElement.classList.contains("dark")
                ? "dark"
                : "light";
            }}
          >
            <svg className="w-5 h-5 mr-2" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
              <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z" />
            </svg>
            ダーク/ライトモード切替
          </button>
        </div>
      </div>
    </div>
  );
}
