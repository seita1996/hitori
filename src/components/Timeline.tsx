import { motion, AnimatePresence } from "framer-motion";
import { Post } from "../types";

interface TimelineProps {
  posts: Post[];
  onDelete?: (id: string) => void;
}

export function Timeline({ posts, onDelete }: TimelineProps) {
  // 日付フォーマット用関数
  const formatDate = (date: Date) => {
    return new Date(date).toLocaleString("ja-JP", {
      year: "numeric",
      month: "numeric",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit"
    });
  };

  // 投稿を古い順に並べ替え（反転）
  const sortedPosts = [...posts].reverse();

  return (
    <div className="space-y-5">
      {posts.length === 0 ? (
        <div className="text-center py-12 px-6 bg-white dark:bg-gray-800 rounded-2xl shadow-md text-gray-500 dark:text-gray-400">
          <svg className="w-12 h-12 mx-auto mb-4 text-gray-300 dark:text-gray-600" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z" />
          </svg>
          <p className="text-lg">まだ投稿がありません。</p>
          <p className="mt-2">思いつくままに何か書いてみましょう。</p>
        </div>
      ) : (
        <AnimatePresence mode="popLayout">
          {sortedPosts.map((post) => (
            <motion.div
              key={post.id}
              initial={{ opacity: 0, y: 20 }}
              animate={{ opacity: 1, y: 0 }}
              exit={{ opacity: 0, x: -100 }}
              transition={{ duration: 0.3, type: "spring", stiffness: 100 }}
              className="bg-white dark:bg-gray-800 rounded-2xl shadow-md p-5 border border-gray-100 dark:border-gray-700"
            >
              <p className="mb-3 whitespace-pre-wrap text-gray-700 dark:text-gray-300 leading-relaxed">{post.content}</p>
              <div className="flex justify-between items-center text-sm text-gray-500 dark:text-gray-400 pt-2 border-t border-gray-100 dark:border-gray-700">
                <span className="flex items-center">
                  <svg className="w-4 h-4 mr-1" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0118 0z" />
                  </svg>
                  {formatDate(post.createdAt)}
                </span>
                {onDelete && (
                  <motion.button
                    whileHover={{ scale: 1.1 }}
                    whileTap={{ scale: 0.9 }}
                    onClick={() => onDelete(post.id)}
                    className="text-red-400 hover:text-red-500 dark:text-red-400 dark:hover:text-red-300 transition-colors duration-200 flex items-center"
                  >
                    <svg className="w-4 h-4 mr-1" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                      <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16" />
                    </svg>
                    削除
                  </motion.button>
                )}
              </div>
            </motion.div>
          ))}
        </AnimatePresence>
      )}
    </div>
  );
}
