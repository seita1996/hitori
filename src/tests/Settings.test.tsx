import { render } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { Settings } from '../components/Settings';

// Mock framer-motion
vi.mock('framer-motion', () => ({
  motion: {
    div: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
    button: ({ children }: { children: React.ReactNode }) => <button>{children}</button>,
    span: ({ children }: { children: React.ReactNode }) => <span>{children}</span>,
  },
}));

describe('Settings', () => {
  it('renders the component', () => {
    const onProviderChangeMock = vi.fn();
    const onSyncMock = vi.fn(() => Promise.resolve());

    render(
      <Settings
        cloudProvider="none" // 例として "none" を渡す
        onProviderChange={onProviderChangeMock}
        onSync={onSyncMock}
        syncStatus="synced" // 例として "synced" を渡す
      />
    );

    // Add assertions here based on the content and behavior of Settings
    // For example, check for settings options or input fields:
    // expect(screen.getByLabelText('Cloud Provider')).toBeInTheDocument();

    // Add snapshot test
    expect(render(
      <Settings
        cloudProvider="none"
        onProviderChange={onProviderChangeMock}
        onSync={onSyncMock}
        syncStatus="synced"
      />
    ).container).toMatchSnapshot();
  });
});
