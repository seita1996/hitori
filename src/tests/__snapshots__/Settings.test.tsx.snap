// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings > renders the component 1`] = `
<div>
  <div
    class="bg-white dark:bg-gray-800 rounded-2xl shadow-md p-6"
  >
    <h2
      class="text-xl font-semibold mb-6 text-gray-800 dark:text-gray-200"
    >
      設定
    </h2>
    <div
      class="space-y-8"
    >
      <div
        class="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700"
      >
        <h3
          class="text-lg font-medium mb-4 text-gray-700 dark:text-gray-300"
        >
          クラウド同期
        </h3>
        <div
          class="space-y-3"
        >
          <label
            class="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors"
          >
            <input
              class="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              type="radio"
            />
            <span
              class="text-gray-700 dark:text-gray-300"
            >
              同期しない
            </span>
          </label>
          <label
            class="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors"
          >
            <input
              class="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              type="radio"
            />
            <span
              class="text-gray-700 dark:text-gray-300 flex items-center"
            >
              Google Drive
            </span>
          </label>
          <label
            class="flex items-center space-x-3 p-2 hover:bg-gray-50 dark:hover:bg-gray-700 rounded-lg cursor-pointer transition-colors"
          >
            <input
              class="h-5 w-5 text-indigo-500 focus:ring-indigo-200"
              type="radio"
            />
            <span
              class="text-gray-700 dark:text-gray-300 flex items-center"
            >
              iCloud
            </span>
          </label>
        </div>
      </div>
      <div
        class="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700"
        style="opacity: 0; height: 0px;"
      >
        <button
          class="px-5 py-2.5 rounded-xl font-medium transition-all duration-200 bg-indigo-500 text-white hover:bg-indigo-600 shadow-sm"
          tabindex="0"
        >
          <span
            class="flex items-center"
          >
            <svg
              class="w-5 h-5 mr-2 "
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            今すぐ同期
          </span>
        </button>
        <div
          class="mt-3 text-sm"
        />
      </div>
      <div
        class="bg-white dark:bg-gray-800 rounded-xl p-5 shadow-sm border border-gray-100 dark:border-gray-700"
      >
        <h3
          class="text-lg font-medium mb-4 text-gray-700 dark:text-gray-300"
        >
          テーマ
        </h3>
        <button
          class="px-5 py-2.5 bg-gray-100 hover:bg-gray-200 dark:bg-gray-700 dark:hover:bg-gray-600 text-gray-700 dark:text-gray-300 rounded-xl transition-colors duration-200 flex items-center"
        >
          <svg
            class="w-5 h-5 mr-2"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
          ダーク/ライトモード切替
        </button>
      </div>
    </div>
  </div>
</div>
`;
